bigquery/google/cloud/bigquery/_pandas_helpers.py

```diff
@@ -285,7 +318,73 @@ def dataframe_to_bq_schema(dataframe, bq_schema):
                 bq_schema_unused
             )
         )
-    return tuple(bq_schema_out)
+
+    # If schema detection was not successful for all columns, also try with
+    # pyarrow, if available.
+    if unknown_type_fields:
+        if not pyarrow:
+            msg = u"Could not determine the type of columns: {}".format(
+                ", ".join(field.name for field in unknown_type_fields)
+            )
+            warnings.warn(msg)
+            return None  # We cannot detect the schema in full.
+
+        # The currate_schema() helper itself will also issue unknown type
+        # warnings if detection still fails for any of the fields.
+        bq_schema_out = currate_schema(dataframe, bq_schema_out)
+
+    return tuple(bq_schema_out) if bq_schema_out else None
+
+
+def currate_schema(dataframe, current_bq_schema):
```

Comment:
I'm not familiar with the verb "currate" (did you mean "curate"?). Augment may be a better verb for what we're doing in this function, as we're adding to the schema. Curate implies removing items.

Reply:
Searching online, it appears that I actually came up with a dubious word, subconsciously trying to score extra points in scrabble or something. :confused: 

Augment indeed sounds better.